### PR TITLE
Cap mblk_madvise() length for errant callers

### PR DESCRIPTION
**Pull request by tristan957**
Signed-off-by: Tristan Partin <tpartin@micron.com>

### CODE REVIEW

File: lib/cn/kvs_mblk_desc.c [hunk 1]
```diff
@@ -78,8 +78,13 @@ mblk_madvise(const struct kvs_mblk_desc *md, size_t off, size_t len, int advice)
     if (len == 0)
         return 0;
 
-    assert(off + len <= md->wlen_pages * PAGE_SIZE);
     assert(md->map_base);
+    assert(off < md->wlen_pages * PAGE_SIZE);
+
+    if (off + len > md->wlen_pages * PAGE_SIZE)
+        len = md->wlen_pages * PAGE_SIZE - off;
```
Comment (tristan957) [inline lib/cn/kvs_mblk_desc.c:85]: After 3.0, I would like to change this to an abort() for IMP replica set test, and see if anybody hits it.

@bhaveshvasandani

File: lib/cn/kvs_mblk_desc.c [hunk 2]
```diff
@@ -78,8 +78,13 @@ mblk_madvise(const struct kvs_mblk_desc *md, size_t off, size_t len, int advice)
...
+
+    assert(off + len <= md->wlen_pages * PAGE_SIZE);
```
Comment (alexttx) [inline lib/cn/kvs_mblk_desc.c:87]: this assert feels unnecessary since it exactly matches the preceding `if` condition.

**Review by alexttx — APPROVED** [review]

File: lib/cn/kvs_mblk_desc.c [hunk 3]
```diff
@@ -78,8 +78,13 @@ mblk_madvise(const struct kvs_mblk_desc *md, size_t off, size_t len, int advice)
     if (len == 0)
         return 0;
 
-    assert(off + len <= md->wlen_pages * PAGE_SIZE);
     assert(md->map_base);
+    assert(off < md->wlen_pages * PAGE_SIZE);
```
Comment (nabeelmmd) [inline lib/cn/kvs_mblk_desc.c:82]: Store md->wlen_pages * PAGE_SIZE in a local variable?

**Review by nabeelmmd — APPROVED** [review]

File: lib/cn/kvs_mblk_desc.c (hunk 1 above)
Comment (beckerg) [reply lib/cn/kvs_mblk_desc.c:86]: I would actually make the off check a live check:
if (len == 0 || off >= wlen)
    return 0;

**Review by beckerg — APPROVED** [review]

File: lib/cn/kvs_mblk_desc.c (hunk 1 above)
Comment (tristan957) [reply lib/cn/kvs_mblk_desc.c:86]: If someone is passing an offset greater than the length of the block, we have pretty large issues and shouldn't ignore it

Comment (beckerg) [reply lib/cn/kvs_mblk_desc.c]: Fair point, but you also shouldn't use it in a release build if it's out of bounds.  Depending upon the advice things could go awry in a very bad way...

Comment (tristan957) [reply lib/cn/kvs_mblk_desc.c]: This goes back to discussion after discussion we've had of assert() vs EBUG. cc @alexttx

Comment (nabeelmmd) [reply lib/cn/kvs_mblk_desc.c]: Agree with Greg on making off >= wlen a live check. I would return EINVAL if this check fails and leave it up to the caller to handle it.

File: lib/cn/kvs_mblk_desc.c [hunk 4]
```diff
@@ -74,12 +74,18 @@ merr_t
...
+    const size_t wlen = md->wlen_pages * PAGE_SIZE;
+
+    assert(md->map_base);
 
     if (len == 0)
         return 0;
 
-    assert(off + len <= md->wlen_pages * PAGE_SIZE);
-    assert(md->map_base);
+    if (off < wlen)
```
Comment (nabeelmmd) [inline lib/cn/kvs_mblk_desc.c:84]: I think you meant to write:
if (off >= wlen)

Comment (tristan957) [reply lib/cn/kvs_mblk_desc.c]: thanks

**Pull request merged**